k8s_dp/k8s_dp.p4

```diff
@@ -915,74 +980,10 @@ control k8s_dp_control(
-                                    meta.src_port = hdr.tcp.src_port;
-                                    meta.dst_port = hdr.tcp.dst_port;
-                                    if (pinned_flows_reverse.apply().miss) {
-                                            pinned_flows.apply();
-                                    }
-                            }
-                    } else {
-                            if (IS_IPV4_UDP) {
-                                    if (tx_balance_udp.apply().hit) {
-                                            create_reverse_ct = true;
-                                            do_clb_pinned_flows_add_on_miss = true;
-                                            //save_to_meta_udp(hdr, meta);
-                                            meta.src_ip = hdr.ipv4.src_addr;
-                                            meta.dst_ip = hdr.ipv4.dst_addr;
-                                            meta.src_port = hdr.udp.src_port;
-                                            meta.dst_port = hdr.udp.dst_port;
-                                            pinned_flows.apply();
-                                    } else {
-                                            //save_to_meta_udp(hdr, meta);
-                                            meta.src_ip = hdr.ipv4.src_addr;
-                                            meta.dst_ip = hdr.ipv4.dst_addr;
-                                            meta.src_port = hdr.udp.src_port;
-                                            meta.dst_port = hdr.udp.dst_port;
-                                            pinned_flows_reverse.apply();
-                                    }
-                            }
-                    }	
-
-                    /* Perform the SNAT or DNAT if enabled by above TCP processing */
-                    switch (meta.mod_action) {
-                            WRITE_SRC_IP: {
-                                    write_source_ip_table.apply();
-                            }
-
-                            WRITE_DEST_IP: {
-                                    write_dest_ip_table.apply();
-                                    if (create_reverse_ct) {
-                                            if (IS_IPV4_TCP) {
-                                                    set_meta_tcp.apply();
-                                            } else {
-                                                    if (IS_IPV4_UDP) {
-                                                            set_meta_udp.apply();
-                                                    }
-                                            }
-                                            pinned_flows_reverse.apply();		
-                                    }
-                            }
-
-                        default: {
-                        }
+                    if ((meta.mod_action == WRITE_SRC_IP) && (meta.nat_done == false)) {
+                        write_source_ip_table.apply();
```

Comment:
SNAT needs to be performed in ingress direction (NET_TO_HOST) (both SNAT and DNAT are performed on client side). Also, it needs to be done after applying ingress policy.

Reply:
Yes, the relative positioning of ingress ACL w.r.t. SNAT is TBD. This commit only fixes the egress ACL to be placed after DNAT.

Reply:
Missed the title of your PR :)
But I thought we are in agreement that we cannot apply ACL on VIP. If that's the case, then ACL has to be applied before SNAT since after SNAT, the source IP will be VIP.
Irrespective of the relative order, SNAT needs to be applied in NET_TO_HOST direction.

Reply:
Yes, we are in agreement --- SNAT should be after ingress ACL and in NET_TO_HOST direction.
However, in this commit, the SNAT is same as before. Both aspects will be fixed as a separate commit.

Reply:
Ok, separate commit is fine, but it will be best to do both in the same PR otherwise egress and ingress will be inconsistent in this version of the code. Egress assumes actual endpoint IP for ACL, while ingress assumes VIP.

Reply:
Yes, it would be nice to do both at the same time. However, that requires untangling DNAT/SNAT and placing them in two different directions. So, instead of submitting one large commit that spans both Policy and Service, we can take incremental commits. At each commit, we ensure that everything remains functional as before while fixing one issue at a time. Without this commit, both ingress and egress Policy is defective where VIP is used instead of pod IP. This commit fixes this issue for egress direction. The fix was ingress direction would come as a separate commit.